tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,213 @@
+// Since buildsys is executed from the context of many different package directories,
+// managing a temporary file via this Rust module
+// prevents having to aquire the path of some static script file on the host system.
+const GO_MOD_SCRIPT: &str = r###"#!/bin/bash
+
+set -e
+
+toplevel=$(tar tf "${LOCAL_FILE_NAME}" | head -1)
+if [ -z "${MOD_DIR}" ] ; then
+    targetdir="${toplevel}"
+else
+    targetdir="${MOD_DIR}"
+fi
+
+tar xf "${LOCAL_FILE_NAME}"
+
+pushd "${targetdir}"
+	go list -mod=readonly ./... >/dev/null && go mod vendor
+popd
+
+tar czf "${OUTPUT}" "${targetdir}"/vendor
+rm -rf "${targetdir}"
+"###;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        external_file: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&external_file.url)?;
+        let local_file_name = &external_file.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(
+            local_file_name.components().count() == 1,
+            error::InputFileSnafu
+        );
+
+        let full_path = package_dir.join(local_file_name);
+        ensure!(
+            full_path.is_file(),
+            error::InputFileBadSnafu { path: full_path }
+        );
+
+        // If a module directory was not provided, set as an empty path.
+        // By default, without a provided module directory, tar will be passed
+        // the first directory found in the archives as the top level Go module
+        let default_empty_path = PathBuf::from("");
+        let mod_dir = external_file
+            .bundle_root_path
+            .as_ref()
+            .unwrap_or(&default_empty_path);
+
+        // Use a default "bundle-{name-of-tar-file}" if no output path was provided
+        let default_output_path =
+            PathBuf::from(format!("bundled-{}", local_file_name.to_string_lossy()));
+        let output_path_arg = external_file
+            .bundle_output_path
+            .as_ref()
+            .unwrap_or(&default_output_path);
+        println!(
+            "cargo:rerun-if-changed={}",
+            output_path_arg.to_string_lossy()
+        );
+
+        // Our SDK and toolchain are picked by the external `cargo make` invocation.
+        let sdk = env::var("BUILDSYS_SDK_IMAGE").context(error::EnvironmentSnafu {
+            var: "BUILDSYS_SDK_IMAGE",
+        })?;
+
+        let args = DockerGoArgs {
+            module_path: package_dir,
+            sdk_image: sdk,
+            go_mod_cache: &root_dir.join(".gomodcache"),
+            command: format!("./{}", GO_MOD_DOCKER_SCRIPT_NAME),
+            extra_docker_args: vec![
+                // Sets container environment variables needed for the docker-go-script
+                format!("-e LOCAL_FILE_NAME={}", local_file_name.to_string_lossy()),
+                format!("-e MOD_DIR={}", mod_dir.to_string_lossy()),
+                format!("-e OUTPUT={}", output_path_arg.to_string_lossy()),
+            ],
```

Comment:
I don't completely love the new "extra docker args" functionality since then it's harder to reason about what the tool will or won't do.

Rather than use variables in the script, it might be worth using placeholders like `__LOCAL_FILE_NAME__` and then just replacing them with these values before writing out the file.

Reply:
Good call - this was likely an implementation artifact from exploring how a script could be used with the `docker-go` tool. I think using a template string is less complex in the end and if someone wanted to look at the script afterwards (or while in process) they could do see those values populated